oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java

```diff
@@ -0,0 +1,455 @@
+   * @throws IOException if the credential cannot be created from the stream
+   */
+  public static ExternalAccountCredentials fromStream(
+      InputStream credentialsStream, HttpTransportFactory transportFactory) throws IOException {
+    checkNotNull(credentialsStream);
+    checkNotNull(transportFactory);
+
+    JsonObjectParser parser = new JsonObjectParser(OAuth2Utils.JSON_FACTORY);
+    GenericJson fileContents =
+        parser.parseAndClose(credentialsStream, StandardCharsets.UTF_8, GenericJson.class);
+    return fromJson(fileContents, transportFactory);
+  }
+
+  /**
+   * Returns external account credentials defined by JSON using the format generated by gCloud.
+   *
+   * @param json a map from the JSON representing the credentials
+   * @param transportFactory HTTP transport factory, creates the transport used to get access tokens
+   * @return the credentials defined by the JSON
+   */
+  static ExternalAccountCredentials fromJson(
+      Map<String, Object> json, HttpTransportFactory transportFactory) {
+    checkNotNull(json);
+    checkNotNull(transportFactory);
+
+    String audience = (String) json.get("audience");
+    String subjectTokenType = (String) json.get("subject_token_type");
+    String tokenUrl = (String) json.get("token_url");
+    String serviceAccountImpersonationUrl = (String) json.get("service_account_impersonation_url");
+
+    Map<String, Object> credentialSourceMap = (Map<String, Object>) json.get("credential_source");
+
+    // Optional params.
+    String tokenInfoUrl =
+        json.containsKey("token_info_url") ? (String) json.get("token_info_url") : null;
+    String clientId = json.containsKey("client_id") ? (String) json.get("client_id") : null;
+    String clientSecret =
+        json.containsKey("client_secret") ? (String) json.get("client_secret") : null;
+    String quotaProjectId =
+        json.containsKey("quota_project_id") ? (String) json.get("quota_project_id") : null;
+
+    if (isAwsCredential(credentialSourceMap)) {
+      return new AwsCredentials(
+          transportFactory,
+          audience,
+          subjectTokenType,
+          tokenUrl,
+          new AwsCredentialSource(credentialSourceMap),
+          tokenInfoUrl,
+          serviceAccountImpersonationUrl,
+          quotaProjectId,
+          clientId,
+          clientSecret,
+          /* scopes= */ null);
+    }
+    return new IdentityPoolCredentials(
+        transportFactory,
+        audience,
+        subjectTokenType,
+        tokenUrl,
+        new IdentityPoolCredentialSource(credentialSourceMap),
+        tokenInfoUrl,
+        serviceAccountImpersonationUrl,
+        quotaProjectId,
+        clientId,
+        clientSecret,
+        /* scopes= */ null);
+  }
+
+  private static boolean isAwsCredential(Map<String, Object> credentialSource) {
+    return credentialSource.containsKey("environment_id")
+        && ((String) credentialSource.get("environment_id")).startsWith("aws");
```

Comment:
possible class cast exception here? I'm not sure, but if so you could catch and return false.